### PR TITLE
[camera] Prevent crash when setting unsupported FocusMode

### PR DESCRIPTION
**Pull request by renefloor**
iOS supports 3 focus modes:
- ContinuousAutoFocus
- AutoFocus
- Fixed
 
Flutter has
- Auto
- Locked

The second (AutoFocus) focuses once and then switches to Fixed.
Previously when setting the mode to 'Auto' the iOS code did check if ContinuousAutoFocus was supported and otherwise set AutoFocus. Setting locked always let iOS set mode to AutoFocus. However, not all devices support AutoFocus, for example the front camera of iPad (5th generation) does not. The FocusMode is always Locked.

My proposed solution is to just not set the FocusMode if it is not supported as that is the same behaviour already done when `ContinuousAutoFocus` is not supported. I can also throw a PlatformException so the caller in the dart side knows that the mode is not supported. However, I don't know if you would also give a warning if `ContinuousAutoFocus` is not supported and you switch to `AutoFocus` instead.

I also need to added a test project as there was none yet, **this test project contains a failing test to check the CI**.

The change is basically this:
from:
```
- (void)applyFocusMode:(FocusMode)focusMode onDevice:(AVCaptureDevice *)captureDevice {
  [captureDevice lockForConfiguration:nil];
  switch (focusMode) {
    case FocusModeLocked:
      [captureDevice setFocusMode:AVCaptureFocusModeAutoFocus];
      break;
    case FocusModeAuto:
      if ([captureDevice isFocusModeSupported:AVCaptureFocusModeContinuousAutoFocus]) {
        [captureDevice setFocusMode:AVCaptureFocusModeContinuousAutoFocus];
      } else {
        [captureDevice setFocusMode:AVCaptureFocusModeAutoFocus];
      }
      break;
  }
  [captureDevice unlockForConfiguration];
}
```
to:
```
- (void)applyFocusMode:(FocusMode)focusMode onDevice:(AVCaptureDevice *)captureDevice {
  [captureDevice lockForConfiguration:nil];
  switch (focusMode) {
    case FocusModeLocked:
      if ([captureDevice isFocusModeSupported:AVCaptureFocusModeAutoFocus]) {
        [captureDevice setFocusMode:AVCaptureFocusModeAutoFocus];
      }
      break;
    case FocusModeAuto:
      if ([captureDevice isFocusModeSupported:AVCaptureFocusModeContinuousAutoFocus]) {
        [captureDevice setFocusMode:AVCaptureFocusModeContinuousAutoFocus];
      } else if ([captureDevice isFocusModeSupported:AVCaptureFocusModeAutoFocus]) {
        [captureDevice setFocusMode:AVCaptureFocusModeAutoFocus];
      }
      break;
  }
  [captureDevice unlockForConfiguration];
}
```

Fixes flutter/flutter#76498

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`. See [plugin_tool format](../script/tool/README.md#format-code))
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Review by mvanbeusekom — COMMENTED** [review]
Looks good in general, I have two nits that would be good to correct.

File: packages/camera/camera/ios/Classes/CameraPlugin.m [hunk 1]
```diff
@@ -994,20 +994,29 @@ - (void)setFocusModeWithResult:(FlutterResult)result mode:(NSString *)modeStr {
...
 
 - (void)applyFocusMode {
-  [_captureDevice lockForConfiguration:nil];
-  switch (_focusMode) {
+  [self applyFocusMode:_focusMode onDevice:_captureDevice];
+}
+
+/// Apply FocusMode on the AVCaptureDevice.
+/// Applies AVCaptureFocusModeAutoFocus when AVCaptureFocusModeContinuousAutoFocus is not supported
+/// for FocusModeAuto Does not apply the focus mode when no mode is supported.
```
[the remark this answers is not in the document]
Comment (renefloor) [reply packages/camera/camera/ios/Classes/CameraPlugin.m]: @mvanbeusekom I applied your suggestion with 2 extra sentences.

**Review by mvanbeusekom — APPROVED** [review]
LGTM!

**Comment by stuartmorgan** [comment]
> I also need to added a test project as there was none yet, **this test project contains a failing test to check the CI**.

Is https://github.com/flutter/plugins/blob/master/packages/camera/camera/ios/Tests/CameraPluginTests.m not being compiled then?

**Comment by renefloor** [comment]
> Is https://github.com/flutter/plugins/blob/master/packages/camera/camera/ios/Tests/CameraPluginTests.m not being compiled then?

Let me check with the cli tools in the plugin repo if it was actually running, but at least it was not visible in the xcode project which makes editing the file way harder.
